### PR TITLE
Add rate limit detection and looping

### PR DESCRIPTION
**Pull request by glensc**
Carries @jbalcorn's https://github.com/Taxel/PlexTraktSync/pull/84:
> Code will detect a rate limit from Trackt and will sleep and loop until the attempted update is complete.

Fixes https://github.com/Taxel/PlexTraktSync/issues/95

Rebased to current `master`

### CODE REVIEW

**Comment by glensc** [comment]
@twolaw did you want to make changes on top of these?
- https://github.com/Taxel/PlexTraktSync/pull/84#discussion_r553220390

you can make the changes here now:
- https://github.com/Taxel/PlexTraktSync/pull/84#discussion_r588757783

**Comment by twolaw** [comment]
> 
> @twolaw did you want to make changes on top of these?


Yes, I'll find time to polish this PR.

**Comment by twolaw** [comment]
@glensc I don't know how to put #129 code inside this PR, I didn't want to create a separated PR.
If you can help me here, thanks.

**Comment by glensc** [comment]
@twolaw can we merge this now?

ps: I'll not review this code, it's too big noise ratio to me, but I can give approval as i see it works for you.

**Review by glensc — COMMENTED** [review]
duh: i can not self-review.

**Review by twolaw — APPROVED** [review]

**Comment by twolaw** [comment]
@glensc tag 0.3.0 ?

**Comment by jbalcorn** [comment]
> Important : the Retry-After value can be read from response only when [this PR](https://github.com/moogar0880/PyTrakt/pull/138) of PyTrakt will be merged. So we should **wait before merging this**.
> 
This explains why I never saw the Retry-After in my testing.

**Comment by glensc** [comment]
@twolaw what is the point of `respect_trakt_rate` / `trakt_post_wait`?

Isn't `Retry-After` the number of seconds from the current time to wait before retry?

If the goal is to set a minimum delay for the wait to be 1.2s, which can be accomplished just `Math.max(header[Retry-After], 1.2)`? am I missing something?

Also, I don't understand the comment about `POST` requests. these do not have `Retry-After` header?

**Comment by glensc** [comment]
@twolaw ah, I see now:

- https://trakt.docs.apiary.io/#introduction/rate-limiting

The limit for POST is super low, 1 call per second, so would immediately hit that limit:

Name | Verb | Methods | Limit
-- | -- | -- | --
AUTHED_API_POST_LIMIT | POST, PUT, DELETE | all | 1 call per second
AUTHED_API_GET_LIMIT | GET | all | 1000 calls every 5 minutes
UNAUTHED_API_GET_LIMIT | GET | all | 1000 calls every 5 minutes

**Comment by twolaw** [comment]
Yes, the point is to avoid rate-limit errors as explained in my previous comment :

> respect POST requests interval to avoid rate-limit errors (instead of going full speed and get many 429 errors)

Maybe it can be improved but I found that `LIMIT + 0.2 sec` is a very good value to get speed and rate-limit respected.

**Comment by glensc** [comment]
@twolaw it's good, I just did not understand it at first.

added the same to the decorator:
- https://github.com/Taxel/PlexTraktSync/pull/177/files

**Comment by twolaw** [comment]
Idea of improvement (not tested) :
We could remove this 1.2 sec hardcoded value (I admit it's not perfect) and go full speed, then at first rate-limit-error raised we could read the Retry-After and set our delay at this value + 0.2
So if one day Trakt change the 1sec value to a lower or higher value, our script would adapt automatically.

File: main.py [hunk 1]
```diff
@@ -72,14 +73,27 @@ def process_movie_section(s, watched_set, ratings_dict, listutil, collection):
...
+                            m.add_to_library()
+                            logging.info('Movie [{} ({})]: Added to trakt collection'.format(
+                                movie.title, movie.year))
+                            break
+                        except trakt.errors.RateLimitException as e:
+                            delay = int(e.response.headers.get("Retry-After", 1))
+                            logging.warning(
+                                "Movie [{} ({})]: Rate Limited on adding to collection. Sleeping {} sec from trakt (GUID: {})".format(movie.title, movie.year, delay, guid))
+                            sleep(delay)
+                            retry += retry
```
Comment (glensc) [inline main.py:93]: what does `retry += retry` supposed to do? seems to be `retry += 0` here as retry is `0` initially, and I don't see where it gets a different value.

Comment (twolaw) [reply main.py:93]: It's a mistake, it should be `retry += 1`
When it reaches 5 it exits the loop.

Comment (glensc) [reply main.py:93]: OK. clear!

Comment (twolaw) [reply main.py:93]: fixed in PR #187

**Pull request merged**